Source/DafnyLanguageServer/Language/VerificationDafnyCodeActionProvider.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Plugins;
+using Microsoft.Dafny.Plugins;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Language;
+
+/// <summary>
+/// A verification quick fixers provides quick "fixes" for verification errors.
+/// For now, it offers to inline a failing postcondition if there is no "return" keyword.
+/// </summary>
+class VerificationDafnyCodeActionProvider : DiagnosticDafnyCodeActionProvider {
+  protected override IEnumerable<DafnyCodeAction>? GetDafnyCodeActions(IDafnyCodeActionInput input, Diagnostic diagnostic, Range selection) {
+    var uri = input.Uri;
+    if (diagnostic.Source != MessageSource.Verifier.ToString()) {
+      return null;
+    }
+
+    if (diagnostic.RelatedInformation?.FirstOrDefault() is not { } relatedInformation) {
+      return null;
+    }
+
+    if (relatedInformation.Location.Uri != uri) {
+      return null;
+    }
+
+    var expression = DafnyCodeActionHelpers.Extract(relatedInformation.Location.Range, input.Code);
```

Comment:
> How come? Code actions are available on lines without diagnostics, are they not?

Not every code action requires extracting some substring of the code all the time. If no code action is returned, or a code action is returned that does not call this particular method, then this line is not called. I won't go into more details as I removed this call.

> This call still seems like an O(N) operation. Why don't you get a DocumentTextBuffer from IDafnyCodeActionInput and let DocumentTextBuffer store a mapping from line to index that can be used to extract text using a Range in constant time?

Well, it's a O(N*C) where C is the number of code actions and N is the size of the file, and C is expected to be very small, so I wouldn't have worried about it.
However, I implemented such a conversion in the implementation  DafnyCodeActionInput, so that the interface can expose the method Extract, and I removed the three ToToken methods that were converting relative positions to absolute positions.